Adding popup_attribute to the GeoJson feature

om-henners opened the pull request:
I've made a change to the GeoJson feature to add a `popup_attribute` parameter. Including this changes the template so that attribute of the feautre is used as the popup for that feature. Based on the map I did for the [#gistribe](http://gistri.be/) 15 minute map competition.

BibMartin commented:
Awesome! Thanks @om-henners .

First reactions :
* I'd have liked to be able to provide `Popup` objects to the function. This can be done in another PR.
  But it would be great if the `popup_attribute` arg were replaced by a `popup_function` (at the cost of having `popup_function = lambda x: 'my_attribute'`) to simplify this second step.

* I would prefer if the `if self.popup_attribute:` were replaced by `{% if this.popup_attribute %}` in the template, so that useser may change their mind and set `popup_attibute` manually after the end of the `__init__` method.

* Having a test would be useful.

* This is really great and would answer @mattwg's question in #375.

om-henners commented:
@BibMartin Happy to put in the full popup support, and you're right - it would be better to have the attribute in a single template. Cheers for the feedback!

om-henners commented:
@BibMartin I've made a coupe of updates to the pull request:

* The `popup_attribute` is now a `popup_function` parameter - if you pass in a function it should take a GeoJson feature and return a string value that is stored in a new `_popupContent` parameter in the feature properties. This is calculated when you intialise the GeoJson class. If you pass a string it acts as before, and simply uses that attribute as the popup value.
* I've put the `{% if this.popup_attribute %}` in the template as requested, so the user can update the `popup_attribute` at a later stage. Since it calculates the results of the function on `__init__` though setting a function here won't work.
* I've added tests for the geojson to test_folium.py
* I've added the popup_function parameter to the `choropleth` method on the map object as well so it will pass straight through

I also took a look at adding the Popup object, but I'll have to have a bit more of a play with it. First though I'd like to update the TopoJson methods so it's consistent with the GeoJson and also has a `popup_function` available.

BibMartin commented on folium/features.py:
```diff
@@ -328,10 +343,36 @@ def style_function(x):
+            self.popup_attribute = popup_function
+        elif isinstance(popup_function, binary_type):
+            self.popup_attribute = text_type(popup_function, 'utf8')
+
```
This is really great ! Thanks.

I would prefer to move this paragraph (346-358) into `style_data` and simply write `self.popup_function = popup_function` here. That way, the following code will also work:

```python
g = GeoJson(data)
g.popup_function = some_function
```

ocefpaf commented:
Wow! Thanks @om-henners!!

@BibMartin Can I assign you to continue reviewing and merge?

om-henners commented:
@ocefpaf no worries - happy to help.

@BibMartin Happy to move the function to the style method. If it's ok I'll delay to tomorrow and do it on the weekend.

BibMartin commented:
> @BibMartin Can I assign you to continue reviewing and merge?

sure you can (and you do) ; I'm looking forward to review this.

> @BibMartin Happy to move the function to the style method. If it's ok I'll delay to tomorrow and do it on the weekend.

No worry, no hurry. This WE is soon enough :wink:

om-henners commented:
I've updated so that the `popup_function` and `popup_attribute` can be set after the fact, and the `popup_function` is called from within the `style_data` method.

om-henners commented:
Hi @BibMartin , any more changes you'd like on this request?

BibMartin commented on folium/features.py:
```diff
@@ -352,6 +410,13 @@ def style_data(self):
+                    feature["properties"]["_popupContent"] = self.popup_function(feature)  # noqa
+            except TypeError:
+                pass
+
```
Why do you need a `try` ?

om-henners replied:
Since the checks on the function are outside the render operation, I was being a bit slack and letting the function fail when it wasn't a function. I'll move checks inside the loop.

BibMartin commented on folium/features.py:
```diff
@@ -328,15 +343,58 @@ def style_function(x):
+        elif isinstance(func, binary_type):
+            self.popup_attribute = text_type(func, 'utf8')
+        self._popup_function = func
+
```
This is great, but it seems complicated to a dummy programmer like me.
Can we simplify this in having a single attibute `popup` (playing the role of both `popup_function` and `popup_attribute` and in testing in the rendering whether it's a function or a string ?
It's a question (I try to keep the code as simple as possible) but your PR is ok anyway.

om-henners replied:
Yeah, no reason not to. I'll update the code. It's easy to move the check into the render function.

BibMartin commented:
Hi @om-henners ,

sorry for the delay ; I've been quite busy this fortnight.

I've put a few question in the code but nothing's blocking. I wait for your reactions before merging.

Thnaks again!

om-henners commented:
Hi @BibMartin , likewise sorry for the delay - been super busy here too :) I've added replies above. Happy to make the updates to the code this weekend.

AlexArcPy commented:
Looking forward to see this functionality merged and pushed into conda package :) I believe there is no existing functionality that would let supply a geoJSON property as a pop-up value, yet this is kind of basic thing for most web mapping operations. @om-henners, brilliant.

ocefpaf commented:
@AlexArcPy thanks for the ping.

@om-henners not sure if I will have time, but if I do are you OK if I pick your commits and finish this PR?

AlexArcPy commented:
@ocefpaf, just realized you are the "ocean guy". Your blog posts are such huge time savers for me, every time I find your posts, I was so glad you spend hours writing them, you are legend.

ocefpaf commented:
> @ocefpaf, just realized you are the "ocean guy". Your blog posts are such huge time savers for me, every time I find your posts, I was so glad you spend hours writing them, you are legend.

:flushed: 

But here on `folium` the "legend" is @BibMartin :wink:

cloud05 commented:
Good day! I would just like to know when this new feature (popup for geojson) would be implemented. So I guess I'll just leave this here to be notified whenever it is ready. Thank you for your work btw. Cheers!

ocefpaf commented:
This one flew under the radar! @om-henners can you rebase so we can move this forward? If not let me know and I will try to pick your commits. Thanks @cloud05 for the ping.

om-henners commented:
@ocefpaf Can do. Sorry, flew under my radar as well! I took a quick look at the rebase - might take a bit of work on features.py, but I should get it done tonight or tomorrow night.

ocefpaf commented:
> @ocefpaf Can do. Sorry, flew under my radar as well! I took a quick look at the rebase - might take a bit of work on features.py, but I should get it done tonight or tomorrow night.

Thanks! There is no rush.

psychemedia commented:
Liking this addition a lot, but looking at `GeoJSON and choropleth.ipynb` I didn't see an example of adding a popup attribute into the `choropleth` method? It may be worth adding one?

cloud05 commented:
Hey @psychemedia, @om-henners already added an example in his first comment. Here it is in a Notebook. Line 16 to be specific.  [http://nbviewer.jupyter.org/github/om-henners/folium/blob/0c2429ce385367f4bf43c59a9d25d980a216c815/examples/GeoJSON%20and%20choropleth.ipynb](url)

psychemedia commented:
@cloud05 I was looking in the `Using choropleth method` section? I saw examples for `folium.GeoJson()` but not `.choropleth()`?

ocefpaf commented:
@om-henners Do you think you can rebase this so we can continue? If not I may try to salvage your commits and issue a new PR.

@BibMartin any final remarks?

cloud05 commented:
Any news about the conflicts in this branch?

MRigal commented:
@om-henners this would be great!

RA-Danny commented:
Is the *popup_function* for *choropleth* added yet ?'

Great worl @om-henners !

psychemedia commented:
Did this feature stall again?

ocefpaf commented:
> Did this feature stall again?

It did. I will rebase this and put into a new PR.

yukclam9 commented:
Hi could I know how we could establish pop up now? I tried the popup_function feature as read from  om-henners 's directory and yet in vain .

jzf2101 commented:
When will this feature be available?

halfdanrump commented:
@jzf2101 @yukclam9 @psychemedia @cloud05 This surely counts as a hack, but making a `folium.GeoJson` object with a `folium.PopUp` for each feature, and then adding those to the map one by one worked fine for me :). It is a bit slow, though.

``` python
import json
import folium
import geopandas as gpd

gdf = gpd.read_file(gpd.datasets.get_path('naturalearth_lowres'))

j = json.loads(gdf.to_json())

geojson = [{'type': j['type'], 'features': [f]} for f in j['features']]

m = folium.Map()

for gj in map(lambda gj: folium.GeoJson(gj), geojson):
    props = gj.data['features'][0]['properties']
    gj.add_child(folium.Popup(str(props)))
    gj.add_to(m)

m
```

ocefpaf commented:
> This surely counts as a hack,

That is actually a nice hack :smile: 

I'll get back to this issue as soon as I finish the re-factoring. I am expand `folium` to take all leaflet's options and use leaflet's defaults.

MRigal commented:
Good luck @ocefpaf ! Don't hesitate to ask for help in this refactoring. It might simply be an open PR with the planned changes as comments/pseudo-code...

pmains commented:
What is the status of this merge?

CSFlorin commented:
I would very much like to know as well. Can we please approve this, fixing the conflicts that are now there of course.

ocefpaf commented:
@CSFlorin `folium` is maintained by volunteers on their free time. So "piling on" an issue won't help.
If you want to try a PR I will gladly review it :wink:

CSFlorin commented:
sure thing, thanks. just seems strange this pull request was never handled when it was new

CSFlorin commented:
side note: couldn't get @halfdanrump 's method to work (map just wouldn't show at all, even when the gdf has just one row), so here's an alternative that just puts a new centered point in each polygon:

    def add_points(mapobj, gdf, popup_field_list):
        # Create empty lists to contain the point coordinates and the point pop-up information
        coords, popups = [], [] 
        # Loop through each record in the GeoDataFrame
        for i, row in gdf.iterrows():
            # Append lat and long coordinates to "coords" list
            coords.append([row.geometry.y, row.geometry.x])
            # Create a string of HTML code used in the IFrame popup
            # Join together the fields in "popup_field_list" with a linebreak between them
            label = '<br>'.join([field + ': ' + str(row[field]) for field in popup_field_list])
            # Append an IFrame that uses the HTML string to the "popups" list 
            popups.append(IFrame(label, width = 300, height = 100))
        
        # Create a Folium feature group for this layer, since we will be displaying multiple layers
        pt_lyr = folium.FeatureGroup(name = 'pt_lyr')
    
        # Add the clustered points of crime locations and popups to this layer
        pt_lyr.add_child(MarkerCluster(locations = coords, popups = popups))
    
        # Add this point layer to the map object
        mapobj.add_child(pt_lyr)
        return mapobj

    gdf['geometry'] = gdf.apply(lambda z: Point(z.geometry.centroid.x, z.geometry.centroid.y), axis=1)
    m = add_points(m, gdf, ['cols whose info you want displayed for that polygon'])
    m

CSFlorin commented:
Does anyone know how to group the layers together so that the popup layer is always above the polygon layer and so that they appear as one in the layercontrol? Much appreciated and sorry for the bugging on this thread

ghandic commented:
This should be solved after GeoJsonCss has been accepted in PR #748 as it allows popups per feature in the GeoJson as well as styling.

MattGraz commented:
I'm a bit unclear as to what the status of this is?  Can someone elaborate a bit on what the holdups are (aside from the obvious conflicts)?

I'm happy to put some time into this and get this feature added to Folium.  Some guidance on whether or not this pull is salvageable and/or I should start from scratch would be appreciated. 

Thanks!

ghandic commented:
This works, have you tried it? We’re just waiting to add a new version to pypi.

Sent from my iPhone

On 21 Jan 2018, at 05:22, MattGraz <notifications@github.com<mailto:notifications@github.com>> wrote:


I'm a bit unclear as to what the status of this is? Can someone elaborate a bit on what the holdups are (aside from the obvious conflicts)?

I'm happy to put some time into this and get this feature added to Folium. Some guidance on whether or not this pull is salvageable and/or I should start from scratch would be appreciated.

Thanks!

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://github.com/python-visualization/folium/pull/376#issuecomment-359224906>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AWaWQX2pQ5burCyoWKGzWtq2PtDZ7sXNks5tMsl8gaJpZM4HlSKV>.

MattGraz commented:
@ghandic: Thanks for the follow up.  I did get it working!  I was just unsure what the status was as far as merging.  Thank you for clarifying!

ritu99 commented:
Any updates?

nittolese commented:
@Conengmo could you resolve the conflicts in merging the pull request?
Thank you!
